uc-crux-llvm/src/UCCrux/LLVM/Postcondition/Type.hs

```diff
@@ -0,0 +1,309 @@
+-- than the value being overwritten, see the module comment.
+data ClobberValue m (inTy :: FullType m) =
+  forall realTy atTy.
+  OpaquePointers realTy ~ OpaquePointers inTy =>
+  ClobberValue
+  { -- | Location of pointer within container value
+    clobberValueCursor :: Cursor m realTy ('FTPtr atTy),
+    -- | Specification of value to write to the pointer
+    clobberValue :: ConstrainedShape m atTy,
+    -- | Type of the container value
+    clobberValueType :: FullTypeRepr m realTy
+  }
+
+data SomeClobberValue m = forall inTy. SomeClobberValue (ClobberValue m inTy)
+
+ppClobberValue :: ClobberValue m inTy -> PP.Doc ann
+ppClobberValue (ClobberValue cur val ty) =
+  PP.hsep
+    [ ppCursor (show (ppConstrainedShape val)) cur
+    , ":"
+    , ppFullTypeRepr ty
+    ]
+
+-- | Specification of how a (part of a) pointer argument is clobbered, i.e.,
+-- which parts to write to and with what data.
+data ClobberArg m (inTy :: FullType m) where
+  DontClobberArg :: ClobberArg m inTy
+  DoClobberArg :: ClobberValue m inTy -> ClobberArg m inTy
+
+data SomeClobberArg m = forall inTy. SomeClobberArg (ClobberArg m inTy)
+
+-- | Untyped postcondition of an LLVM function.
+--
+-- The U stands for untyped, see 'Postcond'.
+--
+-- NOTE(lb): The explicit kind signature here is necessary for GHC 8.8
+-- compatibility.
+data UPostcond m = UPostcond
+  { -- | Specifications of which pointer arguments are written to, and how
+    _uArgClobbers :: IntMap (SomeClobberArg m),
+    -- | Specifications of which global variables are written to, and how
+    _uGlobalClobbers :: Map (GlobalSymbol m) (SomeClobberValue m),
+    -- | Specification of the return value
+    _uReturnValue :: Maybe (ConstrainedTypedValue m)
+  }
+
+uArgClobbers :: Lens' (UPostcond m) (IntMap (SomeClobberArg m))
+uArgClobbers = Lens.lens _uArgClobbers (\s v -> s { _uArgClobbers = v })
+
+uGlobalClobbers :: Lens' (UPostcond m) (Map (GlobalSymbol m) (SomeClobberValue m))
+uGlobalClobbers = Lens.lens _uGlobalClobbers (\s v -> s { _uGlobalClobbers = v })
+
+uReturnValue :: Lens' (UPostcond m) (Maybe (ConstrainedTypedValue m))
+uReturnValue = Lens.lens _uReturnValue (\s v -> s { _uReturnValue = v })
+
+-- | A postcondition which imposes no constraints and has a no return value (is
+-- void).
+emptyUPostcond :: UPostcond m
+emptyUPostcond =
+  UPostcond
+    { _uArgClobbers = IntMap.empty,
+      _uGlobalClobbers = Map.empty,
+      _uReturnValue = Nothing
+    }
+
+-- | A postcondition which imposes no constraints and has some
+-- minimally-constrained return value.
+minimalUPostcond :: ReturnTypeRepr m ft -> UPostcond m
+minimalUPostcond retRepr =
+  UPostcond
+    { _uArgClobbers = IntMap.empty,
+      _uGlobalClobbers = Map.empty,
+      _uReturnValue =
+        case retRepr of
+          FS.VoidRepr -> Nothing
+          FS.NonVoidRepr ft ->
+            Just (ConstrainedTypedValue ft (minimalConstrainedShape ft))
+    }
+
+ppUPostcond :: UPostcond m -> PP.Doc Void
```

Comment:
While we are here, this is another oddly-specific `Void` we could make polymorphic

Reply:
Good catch!